### ISSUE TITLE
ncpa listener and passive not loading

### ISSUE THREAD

**Issue by micoots** [NagiosEnterprises/ncpa#517]
Hi,

I just recently upgraded Debian to the latest packages from buster/sid, and ncpa (v2.0.4 being used) breaks:

# /usr/local/ncpa/ncpa_listener --start
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/cx_Freeze/initscripts/Console.py", line 27, in <module>
  File "ncpa_listener.py", line 5, in <module>
  File "/root/ncpa/agent/ncpadaemon.py", line 14, in <module>
  File "/root/ncpa/agent/listener/database.py", line 5, in <module>
  File "/root/ncpa/agent/listener/server.py", line 11, in <module>
  File "/root/ncpa/agent/listener/psapi.py", line 213, in <module>
  File "/root/ncpa/agent/listener/psapi.py", line 184, in get_root_node
  File "/root/ncpa/agent/listener/psapi.py", line 147, in get_disk_node
  File "/usr/local/lib/python2.7/site-packages/psutil/__init__.py", line 2051, in disk_io_counters
  File "/usr/local/lib/python2.7/site-packages/psutil/_pslinux.py", line 1051, in disk_io_counters
ValueError: not sure how to interpret line '   2       0 fd0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n'

I have tested various incremental updated of the debian packages, up to 2.1.6, and same error results for ncpa_listener and ncpa_passive.

This effectively stops ncpa from passively monitoring the machines I upgraded.

I've spent some hours trouble-shooting this, done various web searches, etc and can't find anything related.

Any ideas on how this can be fixed?

Thanks.

**Comment by bithir** [comment]
It is because you got a kernel update which presents a richer set of values for the psutils - the python libraries used by ncpa does not support this and throw this error when it happens. See https://github.com/giampaolo/psutil/commit/8f99f3782663959062ee868bbfdbc336307a3a4d

There are probably a few options (I'm looking for a solution as well):
- Get the package owner to build ncpa package with a newer version of the python 2.7 libraries
- Update your own system to ensure you have the fix (this might vary pending your distribution) in your system and force ncpa to use those (if this can be done)

**Comment by jomann09** [comment]
I will look into this, building with the new psutils along with an updated python will likely fix the issue.

**Comment by realtebo** [comment]
I am facing probably the same issue on a Debian 9.4 with Python 2.7.13 (all packages upgraded today)

```
# /etc/init.d/ncpa_listener start
Started NCPA Listener
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/cx_Freeze/initscripts/Console.py", line 27, in <module>
  File "ncpa_listener.py", line 5, in <module>
  File "/root/ncpa/agent/ncpadaemon.py", line 14, in <module>
  File "/root/ncpa/agent/listener/database.py", line 5, in <module>
  File "/root/ncpa/agent/listener/server.py", line 11, in <module>
  File "/root/ncpa/agent/listener/psapi.py", line 250, in <module>
  File "/root/ncpa/agent/listener/psapi.py", line 221, in get_root_node
  File "/root/ncpa/agent/listener/psapi.py", line 157, in get_disk_node
  File "/usr/local/lib/python2.7/site-packages/psutil/__init__.py", line 2017, in disk_io_counters
  File "/usr/local/lib/python2.7/site-packages/psutil/_pslinux.py", line 1117, in disk_io_counters
  File "/usr/local/lib/python2.7/site-packages/psutil/_pslinux.py", line 1090, in read_procfs
ValueError: not sure how to interpret line '   1       0 ram0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n'
```

What's not working?

**Comment by jomann09** [comment]
That error looks like the same issue that the user above mentioned in psutils. When we build the next version (2.1.8) it'll include the new version of psutils and should fix this particular issue.

**Comment by realtebo** [comment]
Should I try a build from master? 

I was trying to follows this page but is probably outdated because there is not at all the  `ncpa/build/scripts` folder 

> https://github.com/NagiosEnterprises/ncpa/blob/master/BUILDING.rst#building-on-linux

**Comment by jomann09** [comment]
You could try downgrading it, it may work, you can also try building a new version on a clean system if you want - all you have to do now is clone the repo and run the `build.sh` script, you have to run it as root and it will install packages on the system so doing it on a clean, new system is best.

**Comment by realtebo** [comment]
> You could try downgrading it,

Could you help me, kindly? I downgraded to 2.1.5 but nothing changed.

**Comment by jomann09** [comment]
I can build you an updated version to try, what OS are you using?

**Comment by bithir** [comment]
> > You could try downgrading it,
> 
> Could you help me, kindly? I downgraded to 2.1.5 but nothing changed.

That would not help. The issue is the kernel is incompatible with the python libraries included in the NCPA release.

Your options are:
1) Downgrade kernel
2) Wait for release (see Joman above)
3) Build it cleanly but with patched python libraries/psutil yourself and deploy

**Comment by realtebo** [comment]
I'll chose 2 + 4: "return to to nrpe"

**Comment by bithir** [comment]
> I can build you an updated version to try, what OS are you using?

If you need someone to test, then I could use a fresh build to test - I run Debian Testing (kernel 4.19, I believe) - it is a small system without dependencies, so would be quick to test.

**Comment by bithir** [comment]
Any update on this bug?

**Comment by jomann09** [comment]
I'm going to be building a 2.1.7 probably some time next week, so it will have the new psutils in it.

**Comment by micoots** [comment]
Hi. Do we have a roadmap of NCPA we can look at? or github / bugfix area we could subscribe to?

I'm keen on seeing this issue fixed, and happy to test once there is a fix. 

As more and more systems upgrade kernels, and with debian Buster just around the corner (using kernel 4.19), this fix will become more and more needed.

Much appreciated.

**Comment by jomann09** [comment]
I'm sorry for the wait, this is the next thing to do in my list - once I go through the current issues I'll start building for 2.1.7 - in the meantime if you'd like to try some pre-release builds for testing (they will be missing anything new that gets added before the actual release) of 2.1.7 you can download them here: https://assets.nagios.com/downloads/ncpa/dev/

**Comment by bithir** [comment]
I can confirm that this build solved my problems and that ncpa_listener and ncpa_passive are now running and responding appropriately.

**Comment by jomann09** [comment]
Great, closing this as I am working on building releases.

**Comment by realtebo** [comment]
I too, I confirm the 2.1.7 is working even in my case (and old debian with a new kernel, so a dirty case)

**Comment by Vijay8888** [comment]
@jomann09  is below issue fixed in 2.1.7 and 2.1.8 ncpa versions.
# /etc/init.d/ncpa_listener start
Started NCPA Listener
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/cx_Freeze/initscripts/Console.py", line 27, in <module>
  File "ncpa_listener.py", line 5, in <module>
  File "/root/ncpa/agent/ncpadaemon.py", line 14, in <module>
  File "/root/ncpa/agent/listener/database.py", line 5, in <module>
  File "/root/ncpa/agent/listener/server.py", line 11, in <module>
  File "/root/ncpa/agent/listener/psapi.py", line 250, in <module>
  File "/root/ncpa/agent/listener/psapi.py", line 221, in get_root_node
  File "/root/ncpa/agent/listener/psapi.py", line 157, in get_disk_node
  File "/usr/local/lib/python2.7/site-packages/psutil/__init__.py", line 2017, in disk_io_counters
  File "/usr/local/lib/python2.7/site-packages/psutil/_pslinux.py", line 1117, in disk_io_counters
  File "/usr/local/lib/python2.7/site-packages/psutil/_pslinux.py", line 1090, in read_procfs
ValueError: not sure how to interpret line '   1       0 ram0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n'

**Comment by jomann09** [comment]
I think that might be fixed in the latest version (2.1.8) because it was a bug in the older psutil version that we were using prior.
